tests/tests_deployment/test_keycloak_api.py

```diff
@@ -0,0 +1,1845 @@
+    keycloak_client_id: str,
+    verify_ssl: bool,
+) -> KeycloakAPI:
+    """Create a KeycloakAPI instance without authentication."""
+    return KeycloakAPI(
+        base_url=keycloak_base_url,
+        realm=keycloak_realm,
+        client_id=keycloak_client_id,
+        verify_ssl=verify_ssl,
+    )
+
+
+@pytest.fixture(scope="session")
+def authenticated_keycloak_api(
+    keycloak_base_url: str,
+    keycloak_username: str,
+    keycloak_password: str,
+    keycloak_realm: str,
+    keycloak_client_id: str,
+    verify_ssl: bool,
+) -> KeycloakAPI:
+    """Create an authenticated KeycloakAPI instance for admin operations."""
+    api = KeycloakAPI(
+        base_url=keycloak_base_url,
+        realm=keycloak_realm,
+        client_id=keycloak_client_id,
+        username=keycloak_username,
+        password=keycloak_password,
+        verify_ssl=verify_ssl,
+    )
+    return api
+
+
+@pytest.fixture
+def test_username() -> str:
+    """Generate a unique test username."""
+    return f"testuser_{uuid.uuid4().hex[:8]}"
+
+
+@pytest.fixture
+def test_client_id() -> str:
+    """Generate a unique test client ID."""
+    return f"test-client-{uuid.uuid4().hex[:8]}"
+
+
+@pytest.fixture
+def test_role_name() -> str:
+    """Generate a unique test role name."""
+    return f"test-role-{uuid.uuid4().hex[:8]}"
+
+
+@pytest.fixture
+def test_group_name() -> str:
+    """Generate a unique test group name."""
+    return f"test-group-{uuid.uuid4().hex[:8]}"
+
+
+@pytest.mark.keycloak
+@pytest.mark.filterwarnings("ignore::urllib3.exceptions.InsecureRequestWarning")
+def test_keycloak_login_with_credentials(
+    keycloak_api: KeycloakAPI,
+    keycloak_username: str,
+    keycloak_password: str,
+) -> None:
+    """Test that we can authenticate with Keycloak using username/password."""
+    # Authenticate using KeycloakAPI
+    token_data = keycloak_api.authenticate(keycloak_username, keycloak_password)
+
+    # Verify the response contains expected fields
+    assert "access_token" in token_data
+    assert "refresh_token" in token_data
+    assert "token_type" in token_data
+    assert token_data["token_type"].lower() == "bearer"
+    assert "expires_in" in token_data
+
+    # Verify tokens are stored in the instance
+    assert keycloak_api.access_token is not None
+    assert keycloak_api.refresh_token is not None
+    assert keycloak_api.token_type == "Bearer"
+
+
+@pytest.mark.keycloak
+@pytest.mark.filterwarnings("ignore::urllib3.exceptions.InsecureRequestWarning")
+def test_keycloak_login_invalid_credentials(
+    keycloak_api: KeycloakAPI,
+) -> None:
+    """Test that authentication fails with invalid credentials."""
+    # Authenticate with invalid credentials should raise an HTTPError
+    with pytest.raises(requests.exceptions.HTTPError) as exc_info:
+        keycloak_api.authenticate("invalid_user", "invalid_password")
+
+    # Verify it's a 401 Unauthorized error
+    assert exc_info.value.response.status_code == 401
```

Comment:
For all the cases where we are using numbers for http status codes, let's switch them to actual http.HTTPStatus objects
`assert exc_info.value.response.status_code == HTTPStatus.UNAUTHORIZED` for example. That will make the code clearer without needing to look up what a 401 is